pkg/check/permissions.go

```diff
@@ -0,0 +1,326 @@
+	}
+	return nil
+}
+
+func getFolderReference(ctx context.Context, path string, finder *find.Finder) (*types.ManagedObjectReference, error) {
+	folderObj, err := finder.Folder(ctx, path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "specified folder not found")
+	}
+	ref := folderObj.Reference()
+	return &ref, nil
+}
+
+// CheckAccountPermissions will attempt to validate that the necessary credentials are held by the account performing the
+// installation. each group of privileges will be checked for missing privileges.
+func CheckAccountPermissions(ctx *CheckContext) error {
+	var errs []error
+	err := checkDatastorePrivileges(ctx, ctx.VMConfig.Workspace.DefaultDatastore)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	err = checkDatacenterPrivileges(ctx, ctx.VMConfig.Workspace.Datacenter)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	err = checkFolderPrivileges(ctx, "/", permissionVcenter)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	if ctx.VMConfig.Workspace.Folder != "" {
+		err = checkFolderPrivileges(ctx, ctx.VMConfig.Workspace.Folder, permissionFolder)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return join(errs)
+	}
+	return nil
+}
+
+// CheckNodeAccountPrivileges confirms that resources associated with the node maintain required privileges.
+type CheckNodeAccountPrivileges struct {
+	computeClusters map[string]*mo.ClusterComputeResource
+}
+
+var _ NodeCheck = &CheckNodeAccountPrivileges{}
+
+func (c *CheckNodeAccountPrivileges) Name() string {
+	return "CheckNodeAccountPrivileges"
+}
+
+func (c *CheckNodeAccountPrivileges) StartCheck() error {
+	c.computeClusters = make(map[string]*mo.ClusterComputeResource)
+	return nil
+}
+
+func (c *CheckNodeAccountPrivileges) checkComputeClusterPrivileges(ctx *CheckContext, vm *mo.VirtualMachine) error {
+	cluster, err := getComputeCluster(ctx, vm.Runtime.Host.Reference())
+	if err != nil {
+		klog.Infof("compute cluster resource could not be obtained for %v", vm.Reference())
+		return nil
+	}
+	if _, ok := c.computeClusters[cluster.Name]; ok {
+		klog.Infof("privileges for compute cluster %v have already been checked", cluster.Name)
+		return nil
+	}
+	c.computeClusters[cluster.Name] = cluster
+
+	if _, ok := ctx.VMConfig.VirtualCenter[ctx.VMConfig.Workspace.VCenterIP]; !ok {
+		return errors.New("vcenter instance not found in the virtual center map")
+	}
+	username := ctx.VMConfig.VirtualCenter[ctx.VMConfig.Workspace.VCenterIP].User
+
+	if err := comparePrivileges(ctx.Context, username, cluster.Reference(), ctx.AuthManager, permissions[permissionCluster]); err != nil {
+		return fmt.Errorf("missing privileges for compute cluster %s: %s", cluster.Name, err.Error())
+	}
+	return nil
+}
+
+func (c *CheckNodeAccountPrivileges) CheckNode(ctx *CheckContext, node *v1.Node, vm *mo.VirtualMachine) error {
+	var errs []error
+	err := c.checkComputeClusterPrivileges(ctx, vm)
```

Comment:
Does this needs to be a node check? It seems like permissions are more like cluster scoped and we would be running these checks for each node in the cluster, which seems unnecessary.

Reply:
The node was a reasonably convenient way of finding the compute cluster.  There isn't really a good way to get the compute cluster associated with an OpenShift cluster other than checking an object's parent.  I don't think checking the ComputeCluster privileges are necessarily critical to an already running cluster anyway(outside of maybe the machine API for machine scaling).  I'll remove this NodeCheck and we can revisit if we deem it beneficial to check the privileges on the ComputeCluster.

Reply:
@gnufied I pushed an update to move the cluster privilege check to it's own file.  The check will only check a given compute cluster once for the duration of a given check.  This will provide support for multi-zone clusters when the time comes.